src/extra/components/timepix.py

```diff
@@ -0,0 +1,612 @@
+            hits_tot[dest_slice] = raw['data.tot'][sorted_idx]
+
+            if hits_label is not None:
+                centroids = data[self._centroids_instrument_src.source]
+                hits_label[dest_slice] = centroids['data.labels'][sorted_idx]
+
+        # Load data and bin hits into pulses.
+        psh.map(load_tpx_raw, self._raw_instrument_dc)
+
+        # Build a properly aligned pulse index.
+        index, mask = self._build_pulse_index(
+            tids, pulses, pulse_dim, hits_per_train,
+            hits_pidx, out_of_pulse_events)
+
+        # Build the data frame with detector data and the prepared index.
+        frame_data = {'x': hits_x[mask], 'y': hits_y[mask],
+                      't': hits_tof[mask], 'tot': hits_tot[mask]}
+
+        if extended_columns:
+            frame_data['toa'] = hits_toa[mask]
+            frame_data['pos'] = hits_pos[mask]
+
+            if hits_label is not None:
+                frame_data['label'] = hits_label[mask]
+
+        df = pd.DataFrame(data=frame_data, index=index)
+        df.attrs['pulses'] = pulses
+
+        return df
+
+    def centroid_events(self, pulse_dim='pulseId', toa_offset=0.0,
+                        out_of_pulse_events=False, extended_columns=False,
+                        parallel=None):
+        """Get centroided events as dataframe.
+
+        Loads the processed centroided pixel events as a dataframe and
+        sorts them into pulses using the initialized pulse pattern
+        component. The arguments and behaviour are analogous to
+        [pixel_events()][extra.components.Timepix3.pixel_events].
+
+        Args:
+            pulse_dim ('pulseId' or 'pulseIndex' or 'pulseTime', optional):
+                Label for pulse dimension, pulse ID by default.
+            toa_offset (float, optional): Time-of-arrival offset applied
+                by train in microseconds, 0.0 by default.
+            out_of_pulse_events (bool, optional): Whether to include
+                centroids before the first pulse and after the last
+                pulse in virtual pulses labelled -np.inf and np.inf,
+                False by default. If enabled, the pulse index dimension
+                will be forced to float.
+            extended_columns (bool, optional): Whether to include the
+                original average and maximal time-over-threshold,
+                size and label for each centroid, False by default.
+            parallel (int or None, optional): Nunmber of parallel
+                processes to use, by default 10 or a quarter of all cores
+                whichever is lower. Any non-positive value or 1 disable
+                parallelization.
+
+        Returns:
+            centroids (pandas.DataFrame): Centroids.
+        """
+
+        # Prepare aligned data sources.
+        kd = self.centroids_key.drop_empty_trains()
+        tids = kd.train_ids
+        pulses = self._pulses.select_trains(by_id[tids])
+        pids = pulses.pulse_ids()
```

Comment:
This would be handled in `_sort_timepix_data` through the `KeyError` branch, though this is indistinguishable from not properly having any pulses. We discussed this potential case of swallowing a warning, but it'd require an additional check for train existance via a call like `PulsePattern.pulse_counts()`. As it is now, data of in such a train would be assigned to the trailing pulse, which is easily identifiable and in the worst case misses the data. There is no risk of falsely assigning data.

Reply:
OK, that sounds good. Hopefully missing data from the timeserver is very rare in any case.